### PR TITLE
feat/#1665 Precompile ECRECOVER

### LINKED ISSUE #1665: Precompile: 0x01 ecRecover
### Describe the feature you would like

Scroll has some progress on ecRecover
- https://github.com/scroll-tech/zkevm-circuits/pull/529
- https://github.com/scroll-tech/zkevm-circuits/pull/930

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by KimiWu123**
closed #1665 

This PR was ported from 
- https://github.com/scroll-tech/zkevm-circuits/pull/529
- https://github.com/scroll-tech/zkevm-circuits/pull/930

which manly includes,
1. the main logic in ecrecover gadget (`ecrecover.rs`)
2. signature verifcation circuiit (`sig_circuit.rs`). What I did is to change rlc to word lo/hi.
3. a new table, `sig_table.rs`
4. ecc circuit (`ecc_circuit.rs`). It's not be used in `ecRecover`, but it was implemented in Scroll's PR. If I removed ecc_circuit, it would be inconvienent for people porting `ecAdd`, `ecMul` or `ecPairing`. That's why I keep it here.
5. dependencies update, using `halo2lib` (includes `halo2-base` and `halo2-ecc`)

### CODE REVIEW

File: zkevm-circuits/src/sig_circuit.rs [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+                // This is not optional.
+                let lookup_cells = ecdsa_chip.finalize(&mut ctx);
+                log::info!("total number of lookup cells: {}", lookup_cells);
+
+                ctx.print_stats(&["ECDSA context"]);
+                Ok(assigned_sig_values)
+            },
+        )?;
+
+        layouter.assign_region(
```
Comment (miha-stopar) [inline zkevm-circuits/src/sig_circuit.rs:930]: There are some TODOs in the Scroll version, like `// TODO: is this correct?` before `assign_region` function. I guess it's a leftover?

File: zkevm-circuits/src/ecc_circuit.rs [hunk 2]
```diff
@@ -0,0 +1,1383 @@
...
+        &self,
+        ctx: &mut Context<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        _fr_chip: &FpConfig<F, Fr>,
+        _pairing_chip: &PairingChip<F>,
+        _fp12_chip: &Fp12Chip<F, FpConfig<F, Fq>, Fq12, XI_0>,
+        powers_of_256: &[QuantumCell<F>],
+        op: &EcAddOp,
+    ) -> EcAddDecomposed<F> {
+        log::trace!("[ECC] ==> EcAdd Assignmnet START:");
```
Comment (miha-stopar) [inline zkevm-circuits/src/ecc_circuit.rs:490]: Nitpick: Assignmnet -> Assignment

File: zkevm-circuits/src/ecc_circuit.rs [hunk 3]
```diff
@@ -0,0 +1,1383 @@
...
+        let sum2 = ecc_chip.add_unequal(ctx, &sum1, &point_q, true);
+        let sum2 = ecc_chip.select(ctx, &sum1, &sum2, &point_q_is_zero);
+
+        // sum3 = if R == (0, 0) then sum2 else sum2 - R
+        let sum3 = ecc_chip.sub_unequal(ctx, &sum2, &point_r.ec_point, true);
+        let sum3 = ecc_chip.select(ctx, &sum2, &sum3, &point_r_is_zero);
+
+        ecc_chip.assert_equal(ctx, &rand_point, &sum3);
+
+        log::trace!("[ECC] EcAdd Assignmnet END:");
```
Comment (miha-stopar) [inline zkevm-circuits/src/ecc_circuit.rs:579]: Nitpick: Assignment

File: zkevm-circuits/src/ecc_circuit.rs [hunk 4]
```diff
@@ -0,0 +1,1383 @@
...
+        &self,
+        ctx: &mut Context<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        fr_chip: &FpConfig<F, Fr>,
+        _pairing_chip: &PairingChip<F>,
+        _fp12_chip: &Fp12Chip<F, FpConfig<F, Fq>, Fq12, XI_0>,
+        powers_of_256: &[QuantumCell<F>],
+        op: &EcMulOp,
+    ) -> EcMulDecomposed<F> {
+        log::trace!("[ECC] ==> EcMul Assignmnet START:");
```
Comment (miha-stopar) [inline zkevm-circuits/src/ecc_circuit.rs:611]: Nitpick: Assignment

File: zkevm-circuits/src/ecc_circuit.rs [hunk 5]
```diff
@@ -0,0 +1,1383 @@
...
+        &self,
+        ctx: &mut Context<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        _fr_chip: &FpConfig<F, Fr>,
+        pairing_chip: &PairingChip<F>,
+        fp12_chip: &Fp12Chip<F, FpConfig<F, Fq>, Fq12, XI_0>,
+        powers_of_256: &[QuantumCell<F>],
+        op: &EcPairingOp,
+    ) -> EcPairingDecomposed<F> {
+        log::trace!("[ECC] ==> EcPairing Assignment START:");
```
Comment (miha-stopar) [inline zkevm-circuits/src/ecc_circuit.rs:690]: Nitpick: same as above

**Review by miha-stopar — COMMENTED** [review]
Added some nitpicks and questions. Looks great to me, but I have yet to read ECC circuit more carefully.

File: zkevm-circuits/src/sig_circuit.rs (hunk 1 above)
Comment (KimiWu123) [reply zkevm-circuits/src/sig_circuit.rs:930]: Yep, I believe so. So, I removed it.

File: zkevm-circuits/src/sig_circuit.rs [hunk 6]
```diff
@@ -0,0 +1,1076 @@
...
+        // `cargo test --release --package zkevm-circuits --lib sig_circuit::test::sign_verify --
+        // --nocapture`
+        // - num_advice: 56
+        // - num_lookup_advice: 8
+        // - num_fixed: 1
+        // - lookup_bits: 19
+        // - limb_bits: 88
+        // - num_limbs: 3
+        //
+        // TODO: make those parameters tunable from a config file
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply zkevm-circuits/src/sig_circuit.rs]: fixed in [aed502b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1720/commits/aed502b73465bc38bae1ba69bb83cae3e71d5a3a)

File: zkevm-circuits/src/ecc_circuit.rs (hunk 3 above)
Comment (KimiWu123) [reply zkevm-circuits/src/ecc_circuit.rs:573]: Nice catch! fixed in [4d5070f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1720/commits/4d5070fdf1d7821cab0b4b2caf5f0582fe291ae2)

File: zkevm-circuits/src/ecc_circuit.rs (hunk 4 above)
Comment (KimiWu123) [reply zkevm-circuits/src/ecc_circuit.rs:605]: fixed in [4d5070f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1720/commits/4d5070fdf1d7821cab0b4b2caf5f0582fe291ae2)

File: zkevm-circuits/src/ecc_circuit.rs (hunk 5 above)
Comment (KimiWu123) [reply zkevm-circuits/src/ecc_circuit.rs:684]: fixed in [4d5070f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1720/commits/4d5070fdf1d7821cab0b4b2caf5f0582fe291ae2)

**Review by miha-stopar — APPROVED** [review]
It seems all good to me. I would need some more time to check every detail of the ecc_circuit, but there are tests covering all edge cases and it was peer reviewed by Scroll people, so I think it should be all right.

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs [hunk 7]
```diff
@@ -1614,7 +1619,6 @@ impl<'a> CircuitInputStateRef<'a> {
                         PrecompileCalls::Sha256
                         | PrecompileCalls::Ripemd160
                         | PrecompileCalls::Blake2F
-                        | PrecompileCalls::ECRecover
```
Comment (ChihChengLiang) [inline bus-mapping/src/circuit_input_builder/input_state_ref.rs:1617]: Why do we remove this one here?

File: zkevm-circuits/src/lib.rs [hunk 8]
```diff
@@ -34,6 +34,7 @@ pub mod mpt_circuit;
 pub mod pi_circuit;
 #[cfg(not(target_arch = "wasm32"))]
 pub mod root_circuit;
+pub mod sig_circuit;
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/lib.rs:37]: We should include ecc_circuit here too.
```suggestion
pub mod sig_circuit;
pub mod ecc_circuit;
```

**Review by ChihChengLiang — CHANGES_REQUESTED** [review]
First pass.

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs (hunk 7 above)
Comment (KimiWu123) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs:1617]: The list here is for unsupported precompilers and you can see the code below (L1630).

File: zkevm-circuits/src/lib.rs (hunk 8 above)
Comment (KimiWu123) [reply zkevm-circuits/src/lib.rs:38]: fixed in [5cb6a0e](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1720/commits/5cb6a0e7f7c799bc51f9de8645b83f548174320b)

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs (hunk 7 above)
Comment (ChihChengLiang) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs:1617]: Oh, I see now. Sorry for my false comment.

File: zkevm-circuits/src/lib.rs (hunk 8 above)
Comment (ChihChengLiang) [reply zkevm-circuits/src/lib.rs:37]: I see the reason not to include ecc_circuit module in this PR now. The status quo makes sense to me.

**Review by ChihChengLiang — APPROVED** [review]
LGTM.
